typescript/sdk/src/deploy/verify/BaseContractVerifier.ts

```diff
@@ -0,0 +1,208 @@
+import { MultiProvider } from '../../providers/MultiProvider.js';
+import { ChainName } from '../../types.js';
+
+import {
+  BuildArtifact,
+  ContractVerificationInput,
+  SolidityStandardJsonInput,
+} from './types.js';
+import { FamilyVerificationDelay } from './utils.js';
+
+export abstract class BaseContractVerifier {
+  protected logger = rootLogger.child({ module: this.constructor.name });
+  protected contractSourceMap: { [contractName: string]: string } = {};
+  protected readonly standardInputJson: SolidityStandardJsonInput;
+
+  constructor(
+    protected readonly multiProvider: MultiProvider,
+    buildArtifact: BuildArtifact,
+  ) {
+    this.standardInputJson = buildArtifact.input;
+    this.createContractSourceMapFromBuildArtifacts();
+  }
+
+  protected createContractSourceMapFromBuildArtifacts(): void {
+    const contractRegex = /contract\s+([A-Z][a-zA-Z0-9]*)/g;
+    Object.entries(this.standardInputJson.sources).forEach(
+      ([sourceName, { content }]) => {
+        const matches = content.matchAll(contractRegex);
+        for (const match of matches) {
+          const contractName = match[1];
+          if (contractName) {
+            this.contractSourceMap[contractName] = sourceName;
+          }
+        }
+      },
+    );
+  }
+
+  public async verifyContract(
+    chain: ChainName,
+    input: ContractVerificationInput,
+    logger = this.logger,
+  ): Promise<void> {
+    const verificationLogger = logger.child({
+      chain,
+      name: input.name,
+      address: input.address,
+    });
+
+    if (!this.shouldVerifyContract(chain, input, verificationLogger)) {
+      return;
+    }
+
+    const explorerApi = this.multiProvider.tryGetExplorerApi(chain);
+
+    await sleep(
+      FamilyVerificationDelay[
+        explorerApi?.family as keyof typeof FamilyVerificationDelay
+      ] ?? 0,
+    );
+    await this.verify(chain, input, verificationLogger);
+  }
+
+  protected shouldVerifyContract(
+    chain: ChainName,
+    input: ContractVerificationInput,
+    verificationLogger: Logger,
+  ): boolean {
+    const metadata = this.multiProvider.tryGetChainMetadata(chain);
+    const rpcUrl = metadata?.rpcUrls[0].http ?? '';
+    if (rpcUrl.includes('localhost') || rpcUrl.includes('127.0.0.1')) {
+      verificationLogger.debug('Skipping verification for local endpoints');
+      return false;
+    }
+
+    const explorerApi = this.multiProvider.tryGetExplorerApi(chain);
+    if (!explorerApi) {
+      verificationLogger.debug('No explorer API set, skipping');
+      return false;
+    }
+
+    if (!explorerApi.family) {
+      verificationLogger.debug(`No explorer family set, skipping`);
+      return false;
+    }
+
+    if (explorerApi.family === ExplorerFamily.Other) {
+      verificationLogger.debug(`Unsupported explorer family, skipping`);
+      return false;
+    }
+
+    if (input.address === ethers.constants.AddressZero) return false;
```

Comment:
see `isZeroishAddress` in the utils package

Reply:
Done!